src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -0,0 +1,1533 @@
+            {
+                return nullptr;
+            }
+
+            return NewExtension(cast->IsUnsigned() ? ScevOper::ZeroExtend : ScevOper::SignExtend, TYP_LONG, op);
+        }
+        case GT_ADD:
+        case GT_MUL:
+        case GT_LSH:
+        {
+            Scev* op1 = Analyze(block, tree->gtGetOp1(), depth + 1);
+            if (op1 == nullptr)
+                return nullptr;
+
+            Scev* op2 = Analyze(block, tree->gtGetOp2(), depth + 1);
+            if (op2 == nullptr)
+                return nullptr;
+
+            ScevOper oper;
+            switch (tree->OperGet())
+            {
+                case GT_ADD:
+                    oper = ScevOper::Add;
+                    break;
+                case GT_MUL:
+                    oper = ScevOper::Mul;
+                    break;
+                case GT_LSH:
+                    oper = ScevOper::Lsh;
+                    break;
+                default:
+                    unreached();
+            }
+
+            return NewBinop(oper, op1, op2);
+        }
+        case GT_COMMA:
+        {
+            return Analyze(block, tree->gtGetOp2(), depth + 1);
+        }
+        case GT_ARR_ADDR:
+        {
+            return Analyze(block, tree->AsArrAddr()->Addr(), depth + 1);
+        }
+        default:
+            return nullptr;
+    }
+}
+
+//------------------------------------------------------------------------
+// CreateSimpleAddRec: Create a "simple" add-recurrence. This handles the most
+// common patterns for primary induction variables where we see a store like
+// "i = i + 1".
+//
+// Parameters:
+//   headerStore  - Phi definition of the candidate primary induction variable
+//   enterScev    - SCEV describing start value of the primary induction variable
+//   stepDefBlock - Block containing the def of the step value
+//   stepDefData  - Value of the def of the step value
+//
+// Returns:
+//   SCEV node if this is a simple addrec shape. Otherwise nullptr.
+//
+Scev* ScalarEvolutionContext::CreateSimpleAddRec(GenTreeLclVarCommon* headerStore,
+                                                 ScevLocal*           enterScev,
+                                                 BasicBlock*          stepDefBlock,
+                                                 GenTree*             stepDefData)
+{
+    if (!stepDefData->OperIs(GT_ADD))
+    {
+        return nullptr;
+    }
+
+    GenTree* stepTree;
+    GenTree* op1 = stepDefData->gtGetOp1();
+    GenTree* op2 = stepDefData->gtGetOp2();
+    if (op1->OperIs(GT_LCL_VAR) && (op1->AsLclVar()->GetLclNum() == headerStore->GetLclNum()) &&
+        (op1->AsLclVar()->GetSsaNum() == headerStore->GetSsaNum()))
+    {
+        stepTree = op2;
+    }
+    else if (op2->OperIs(GT_LCL_VAR) && (op2->AsLclVar()->GetLclNum() == headerStore->GetLclNum()) &&
+             (op2->AsLclVar()->GetSsaNum() == headerStore->GetSsaNum()))
+    {
+        stepTree = op1;
+    }
+    else
+    {
+        return nullptr;
```

Comment:
Maybe add a comment here noting that this null return means we do not have the simple IV shape (1 direct update of the def from the phi)?